gems/aws-sdk-core/lib/aws-sdk-core/client_side_monitoring/request_metrics.rb

```diff
@@ -0,0 +1,129 @@
+module Aws
+  module ClientSideMonitoring
+    class RequestMetrics
+      attr_reader :api_call, :api_call_attempts
+
+      def initialize(opts = {})
+        @service = opts[:service]
+        @api = opts[:operation]
+        @client_id = opts[:client_id]
+        @timestamp = opts[:timestamp] # In epoch milliseconds
+        @version = 1 # Currently hard-coded
+        @api_call = ApiCall.new(@service, @api, @client_id, @version, @timestamp)
+        @api_call_attempts = []
+      end
+
+      def build_call_attempt(opts = {})
+        timestamp = opts[:timestamp]
+        fqdn = opts[:fqdn]
+        region = opts[:region]
+        user_agent = opts[:user_agent]
+        access_key = opts[:access_key]
+        session_token = opts[:session_token]
+        ApiCallAttempt.new(
+          @service,
+          @api,
+          @client_id,
+          @version,
+          timestamp,
+          fqdn,
+          region,
+          user_agent,
+          access_key,
+          session_token
+        )
+      end
+
+      def add_call_attempt(attempt)
+        @api_call_attempts << attempt
+      end
+
+      class ApiCall
+        attr_reader :service, :api, :client_id, :timestamp, :version,
+          :attempt_count, :latency
+
+        def initialize(service, api, client_id, version, timestamp)
+          @service = service
+          @api = api
+          @client_id = client_id
+          @version = version
+          @timestamp = timestamp
+        end
+
+        def complete(opts = {})
+          @latency = opts[:latency]
+          @attempt_count = opts[:attempt_count]
+        end
+
+        def to_json(*a)
+          {
+            "Type" => "ApiCall",
+            "Service" => @service,
+            "Api" => @api,
+            "ClientId" => @client_id,
+            "Timestamp" => @timestamp,
+            "Version" => @version,
+            "AttemptCount" => @attempt_count,
+            "Latency" => @latency
+          }.to_json
+        end
+      end
+
+      class ApiCallAttempt
+        attr_reader :service, :api, :client_id, :version, :timestamp, :fqdn,
+          :region, :user_agent, :access_key, :session_token
+        attr_accessor :request_latency, :http_status_code, :aws_exception_msg,
+          :x_amz_request_id, :x_amz_id_2, :x_amzn_request_id
+
+        def initialize(
+          service,
+          api,
```

Comment:
Lol love to see long list of variable aligned 🎉 Is the concern there trying to make the ApiCall/ApiCallAttempt to be explicit as they are private used classed?

Reply:
I should add the `# @api private` tag, agreed.